src/Visualization.jl

```diff
@@ -28,27 +28,52 @@ function Makie.plot!(f::Makie.GridPosition, tn::TensorNetwork{A}; labels = false
     # TODO recognise them by using `DeltaArray` or `Diagonal` representations
     copytensors = findall(t -> haskey(t.meta, :dual), tensors(tn))
 
-    kwargs[:node_size] = [max(15, log2(size(tensors(tn, i)) |> prod)) for i in 1:nv(graph)]
-    kwargs[:node_marker] = [i ∈ copytensors ? :diamond : :circle for i in 1:length(tensors(tn))]
-    kwargs[:node_color] = [i ∈ copytensors ? :black : :white for i in 1:length(tensors(tn))]
+    oinds = openinds(tn)
+    ghostnodes = []
+    while !isempty(oinds)
+        for ind in oinds
+            tensor = links(ind)[1]
+            tensor_oinds = filter(id -> id.name ∈ tensor.labels, oinds)
+
+            add_vertex!(graph)
+            add_edge!(graph, pos[tensor], nv(graph))
+
+            push!(ghostnodes, nv(graph))
+            filter!(id -> id ∉ tensor_oinds, oinds)
+        end
+    end
+
+    kwargs[:node_size] = [i ∈ ghostnodes ? 0 : max(15, log2(size(tensors(tn, i)) |> prod)) for i in 1:nv(graph)]
+    kwargs[:node_marker] = [i ∈ copytensors ? :diamond : :circle for i in 1:nv(graph)]
+    kwargs[:node_color] = [i ∈ copytensors ? :black : :white for i in 1:nv(graph)]
 
     if labels
         elabels = Vector{String}([])
         elabels_color = Vector{Symbol}([])
 
         for edge in edges(graph)
             copies = filter((x -> x ∈ copytensors), [edge.src, edge.dst])
+            notghosts = filter((x -> x ∉ ghostnodes), [edge.src, edge.dst])
 
-            if isempty(copies) # there are no copy tensors in the nodes of this edge
-                push!(elabels, join(Tenet.labels(tensors(tn)[edge.src]) ∩ Tenet.labels(tensors(tn)[edge.dst]), ','))
-                push!(elabels_color, :black)
+            if length(notghosts) == 2 # there are no ghost nodes in this edge
+                if isempty(copies) # there are no copy tensors in the nodes of this edge
+                    push!(elabels, join(Tenet.labels(tensors(tn)[edge.src]) ∩ Tenet.labels(tensors(tn)[edge.dst]), ','))
+                    push!(elabels_color, :black)
+                else
+                    push!(elabels, string(tensors(tn)[copies[]].meta[:dual]))
+                    push!(elabels_color, :grey)
+                end
             else
-                push!(elabels, string(tensors(tn)[copies[]].meta[:dual]))
-                push!(elabels_color, :grey)
+                tensor_oinds = filter(id -> id.name ∈ tensors(tn)[notghosts[1]].labels, openinds(tn))
+                indices = (id -> id.name).(tensor_oinds)
+
+                push!(elabels, join(indices, ","))
+                push!(elabels_color, :black)
```

Comment:
Could you explain me this code? I guess it adds the `labels` to the plot but Im not sure of some steps.

Reply:
This part of the code it checks the two nodes at each vertex. If there are no `ghostnode`s in the nodes, it `join`s the labels. If there is one `ghostnode` (the only other possibility is this one) it takes the label of the `notghost`.

Reply:
I'm gonna restate the same as above. I don't think joining open indices is a good idea. Actually, from watching the plot you should be able to count the amount of open indices but if you join indices, then you cannot or it's much more difficult. Open indices should remain separate.